### DISCUSSION TITLE
basic Basic: a one pass basic Basic compiler for a basic dialect of Basic.

### DISCUSSION THREAD

**Discussion by abdulbahajaj** [maximecb/uvm#15] [Show and tell]
I hacked together a Basic to UVM assembly compiler. Checkout the docs [here](https://github.com/abdulbahajaj/basic-Basic) for more details.

Here is an example code:

```basic
LET red 13583410
LET lightRed 16541523
LET blue 7574225 
LET white 16777215
LET brown 7355966
LET lightBrown 9393750
LET grey 10921638
LET lightGrey 12105912

LET height 800
LET width 800
LET seaLevel (height/4)*3

FOR x 0 TO width-1
FOR y 0 TO height-1
IF y > seaLevel THEN LET color blue ELSE LET color white
PLOT x y color
NEXT y
NEXT x

LET startAt width/20
LET diagonalLen width/6
LET sink 25
LET borderWidth 4

FOR c 0 TO diagonalLen

LET y (seaLevel +  c) - diagonalLen + sink

FOR f startAt + c TO width - (startAt + c) 
PLOT f y brown
NEXT f
LET x startAt + c
FOR depth 0 TO borderWidth
LET ywithDepth y + depth
PLOT x ywithDepth lightBrown
PLOT width - x ywithDepth lightBrown
NEXT depth

NEXT c


LET lowestX (startAt + diagonalLen)
LET y seaLevel + sink
FOR x lowestX TO (width - lowestX)
FOR depth 0 TO borderWidth
PLOT x y + depth   lightBrown
NEXT depth
NEXT x

LET boatTop seaLevel - diagonalLen + sink

FOR x startAt TO (width - startAt)
FOR depth 0 TO borderWidth
PLOT x boatTop-depth lightBrown
NEXT depth
NEXT x

LET poleXStart startAt*4+10
LET poleWidth 5
FOR y height/3-50 TO boatTop-borderWidth-1

FOR depth 0 TO 3
PLOT poleXStart-depth y lightGrey
PLOT poleXStart-depth y lightGrey
NEXT depth

FOR depth 0 TO poleWidth
PLOT poleXStart+depth y grey
NEXT depth
NEXT y

LET flagEnd width/2 + poleXStart 
FOR y height/3-50 TO boatTop-80
FOR x poleXStart + poleWidth + 1 TO flagEnd
PLOT x y red
NEXT x
NEXT y

FOR depth 0 TO 2
FOR x poleXStart + poleWidth + 1 TO flagEnd
PLOT x y+depth lightRed
NEXT x
NEXT depth
```
Will produce the following

![This is an image](https://raw.githubusercontent.com/abdulbahajaj/basic-Basic/main/pirate-ship.png)

Happy to answer any questions

**Comment by maximecb** [comment]
Hi Abdul! That's super cool! :)

I'm impressed that your compiler is so small as well. If you're able to produce a higher resolution screenshot, we could use it to promote UVM (if you're ok with that?) :)

In case this interests you, I've added some primitives for text input. made a simple example into `ncc/examples/textedit.c`. I built this with the idea that maybe it could be used to build something like a basic interpreter / REPL:

<img width="460" alt="Screen Shot 2023-03-08 at 6 39 34 PM" src="https://user-images.githubusercontent.com/713766/223877647-e626f17a-f773-4dd2-b10d-6434c16b7fc1.png">

**Reply by abdulbahajaj** [reply]
Hey Maxime, glad you like it!
By higher resolution do you mean you want the squares that the PLOT command draws to be smaller? (they are more apparent in diagonal lines )

The demo looks cool!

**Reply by maximecb** [reply]
> By higher resolution do you mean you want the squares that the PLOT command draws to be smaller?

Yes, maybe something like a 400x400 window  or 320x240 resolution :)

**Reply by abdulbahajaj** [reply]
I have updated the compiler to generate a window size of 800x800 containing 400x400 pixels
Here is the boat picture updated to reflect the new resolution:

<img width="800" alt="pirate-ship" src="https://user-images.githubusercontent.com/6544882/223930272-0f9df79c-a311-41ff-ac00-bfbf85ca7f4d.png">

Is this what you are looking for?

**Reply by abdulbahajaj** [reply]
Updated it again to be 800x800 window with 800x800 pixels in case you wanted a higher resolution:

<img width="801" alt="Screen Shot 2023-03-09 at 1 22 47 AM" src="https://user-images.githubusercontent.com/6544882/223937850-848f5eca-de21-42ea-a535-c419b1c9e23b.png">

**Reply by maximecb** [reply]
Nice! And you're ok with me including these screenshots in the UVM repo under the Apache-2 license? :)

**Reply by abdulbahajaj** [reply]
Yes

**Reply by maximecb** [reply]
Wonderful. Thank you :)

Any chance I could convince you to work on a basic REPL compiled by the ncc compiler in `uvm/ncc`? :D

**Reply by abdulbahajaj** [reply]
I am thinking about it. Can you tell me more about what you would like it to do? Would there be two windows, one for user to input commands like in the textedit.c file and the other for the output of plot?

**Reply by maximecb** [reply]
At the moment only one window is supported. Though we could change that, I already structured the code so there is a `window_id` value to support multiple windows. If you're interested in helping add support for multiple windows, the code for that is in `vm/src/sys/window.rs`

Otherwise, could structure it with just one window, and either change into a graphics mode, or allow outputting images directly into the REPL.

**Reply by abdulbahajaj** [reply]
I have decided to go ahead with an implementation of a basic REPL in NCC. I have some ideas on how to make it work in 1 window. Stay tuned!

**Reply by maximecb** [reply]
Oh yeah! I'd be happy to merge that into `ncc/examples` if you're able to make it self-contained in one C file and share it under a CC0 license :)

**Comment by maximecb** [comment]
Hi @abdulbahajaj !

I recently resumed work on UVM. When trying to run the basic program, I seem to be getting an error:
```
./build_and_run.sh examples/basic.c 
thread 'main' panicked at 'attempting to access data of type u32 at unaligned address', src/vm.rs:592:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This might be because I added checks to make sure that memory accesses are aligned. Would you have time to investigate the issue? I would like to keep this example working :)

**Reply by abdulbahajaj** [reply]
Hey, yes I will fix it asap.

**Reply by maximecb** [reply]
Wonderful! :)

**Reply by abdulbahajaj** [reply]
Hey @maximecb, Looked at the issue
It looks like it is a malloc problem. If I allocated memory that is not a multiple of sizeof(u32) the next time I allocate memory it fails with this problem.
My guess:
The unaligned access seem to happen inside malloc. When I allocate any amount of memory that is not a multiple of sizeof(u32), `__next_alloc__` stops being u32 aligned. Then the following lines cast its value to u32 and then write to it which is causing the failure.

```c
    // Write a magic word at the beginning of the block for safety checks
    u32* magic_ptr = (u32*)header_ptr;
    *magic_ptr = 0x1337BAB3;
```

For example

```c
#include <stdlib.h>

int main() {
  malloc(4);
  malloc(4);
  malloc(4);
  malloc(4);
  malloc(1);
}
```
succeeds

```c
#include <stdlib.h>

int main() {
  malloc(1);
  malloc(4);
}
```
Fails with 
```thread 'main' panicked at src/vm.rs:592:13:
attempting to access data of type u32 at unaligned address
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Reply by maximecb** [reply]
Thanks for diagnosing this Abdul 🙏  I just pushed a fix :D

**Reply by abdulbahajaj** [reply]
np, had fun investigating it XD
